File: src/main/java/gargoyle/sexbomb/util/applet/GXApplet.java
package gargoyle.sexbomb.util.applet;

import gargoyle.sexbomb.util.res.Resources;

public abstract class GXApplet extends G0Applet {
    protected GXApplet(Resources resources) {
        super(resources);
    }

    protected static void run(Class<? extends GXApplet> clazz, String[] args) {
        _run(clazz, args);
    }

    protected abstract void doDestroy();

    protected abstract void doInit();

    protected abstract void doStart();

    protected abstract void doStop();

    @Override
    public final synchronized void init() {
        setFocusable(true);
        try {
            doInit();
        } catch (RuntimeException e) {
            error(e);
        }
    }

    @Override
    public final synchronized void start() {
        try {
            doStart();
        } catch (RuntimeException e) {
            error(e);
        }
    }

    @Override
    public final synchronized void stop() {
        try {
            doStop();
        } catch (RuntimeException e) {
            error(e);
        }
    }

    @Override
    public final synchronized void destroy() {
        try {
            doDestroy();
        } catch (RuntimeException e) {
            error(e);
        }
    }
}

File: src/main/kotlin/gargoyle/sexbomb/util/applet/GXApplet.kt
package gargoyle.sexbomb.util.applet

import gargoyle.sexbomb.SexBomb
import gargoyle.sexbomb.util.res.Resources
import kotlin.reflect.KClass

abstract class GXApplet protected constructor(resources: Resources) : G0Applet(resources) {
    protected abstract fun doDestroy()
    protected abstract fun doInit()
    protected abstract fun doStart()
    protected abstract fun doStop()
    @Synchronized
    override fun init() {
        isFocusable = true
        try {
            doInit()
        } catch (e: RuntimeException) {
            error(e)
        }
    }

    @Synchronized
    override fun start() {
        try {
            doStart()
        } catch (e: RuntimeException) {
            error(e)
        }
    }

    @Synchronized
    override fun stop() {
        try {
            doStop()
        } catch (e: RuntimeException) {
            error(e)
        }
    }

    @Synchronized
    override fun destroy() {
        try {
            doDestroy()
        } catch (e: RuntimeException) {
            error(e)
        }
    }

    companion object {
        fun run(clazz: KClass<SexBomb>, args: Array<String>) {
            run0(clazz, args)
        }
    }
}
